### PR TITLE
Add perplexity calculation for Tensor and Pipeline parallized Llama models

### PR DESCRIPTION
**Pull request by Alex-Vasile**
Pipelined perplexity test will be enabled seperately and tracked by https://github.com/nod-ai/shark-ai/issues/1306.

### CODE REVIEW

File: sharktank/sharktank/evaluate/perplexity_iree.py [hunk 1]
```diff
@@ -206,24 +225,46 @@ def prefill_vmfb(self, token_batch: torch.tensor, i: int) -> torch.tensor:
...
-            self.cache_state,
+
+        decode_kwargs = OrderedDict(
+            [
+                ("tokens", token_batch),
+                ("seq_lens", [self.batch.seq_lens]),
+                (
+                    "start_positions",
+                    [self.batch.seq_lens.clone()],
+                ),  # TODO: Why clone here and not in prefill
```
Comment (archana-ramalingam) [inline sharktank/sharktank/evaluate/perplexity_iree.py:245]: We do this here because for every decode step, we increase `self.batch.seq_lens` by 1 in line 235 in this script.

File: sharktank/sharktank/evaluate/perplexity_iree.py [hunk 2]
```diff
@@ -235,52 +276,77 @@ def decode_vmfb(self, token_batch: torch.tensor, i: int) -> torch.tensor:
...
-                    decode_logits = self.decode_vmfb(token_batch, i)
-                    self.out_logits.append(decode_logits)
-
-                iter += 1
-
-            out_logits = torch.cat(self.out_logits, dim=1)
+                    decode_logits = self.decode_vmfb(token_batch, i, devices)
+                    out_logits.append(decode_logits)
 
+            out_logits = torch.cat(out_logits, dim=1)
```
Comment (archana-ramalingam) [inline sharktank/sharktank/evaluate/perplexity_iree.py:340]: This can be rewritten as `ops.cat` too.

File: sharktank/tests/evaluate/perplexity_iree_test.py [hunk 3]
```diff
@@ -47,7 +47,6 @@ def test_llama3_8B_f16(self):
             [
                 f"--irpa-file={self.llama3_8b_f16_model}",
                 f"--tokenizer-config-json={self.llama3_8b_tokenizer}",
-                f"--iree-device={self.iree_device}",
```
Comment (archana-ramalingam) [inline sharktank/tests/evaluate/perplexity_iree_test.py:50]: Is there a reason we are removing the `--iree-device` here?

File: sharktank/tests/layers/pipelined_sharded_paged_attention_test.py [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+        cache_state: List[torch.Tensor],
+        pipelined_sharded_cache_state: List[SplitPrimitiveTensor],
+    ):
+        pipelined_sharded_states_as_unsharded = [
+            ops.unshard(unflatted_page).flatten(start_dim=1)
+            for unflatted_page in self.pipelined_sharded_cache.unflatten_page_tables(
+                pipelined_sharded_cache_state
+            )
+        ]
+        pipelined_sharded_states_as_single = torch.cat(
```
Comment (archana-ramalingam) [inline sharktank/tests/layers/pipelined_sharded_paged_attention_test.py:88]: Again change to ops.cat

Comment (archana-ramalingam) [reply sharktank/tests/layers/pipelined_sharded_paged_attention_test.py]: Looks like it's fixed now.

File: sharktank/tests/evaluate/perplexity_iree_test.py [hunk 5]
```diff
@@ -43,16 +48,124 @@ def test_llama3_8B_f16(self):
...
                 f"--iree-hal-target-device={self.iree_hal_target_device}",
                 f"--iree-hip-target={self.iree_hip_target}",
-                f"--tensor-parallelism-size=1",
+                f"--tensor-parallelism-size=2",
+                f"--pipeline-parallelism-size=1",
                 f"--attention-kernel=torch",
                 f"--num-prompts={self.batch_size}",
+                f"--use-attention-mask",
```
Comment (archana-ramalingam) [inline sharktank/tests/evaluate/perplexity_iree_test.py]: Do we need to pass mask explicitly for tp?

**Review by archana-ramalingam — APPROVED** [review]

File: sharktank/tests/evaluate/perplexity_iree_test.py (hunk 5 above)
Comment (Alex-Vasile) [reply sharktank/tests/evaluate/perplexity_iree_test.py]: Yes, required until https://github.com/nod-ai/shark-ai/issues/1202 is fixed.
